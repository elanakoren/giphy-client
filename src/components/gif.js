// @flow
import React, {Component} from 'react';
import GifDetails from './gifDetails';

type Props = {
    imgSrc: string,
    rating: string,
    username: string,
    import_datetime: string,
};

type State = {
    expanded: boolean
};

export class Gif extends Component<Props, State> {
    props: Props;
    state: State;

    constructor(props: Props) {
        super(props);
        this.state = {expanded: false};
    }

    onGifClick() {
        const {expanded} = this.state;
        this.setState({expanded: !expanded});
    }

    // I want every GIF to go back to its unexpanded state if a sort or a search
    // occurs.
    componentWillReceiveProps() {
        this.setState({expanded: false})
    }

    render() {
        const {imgSrc, rating, username, import_datetime} = this.props;
        const {expanded} = this.state;
        let showGifDetails = expanded ? 'show-details' : 'hide-details';
        let expandGif = expanded ? 'expanded' : '';
        return (
            <div className="grid-cell">
                <img
                    alt=""
                    className={`gif ${expandGif}`}
                    src={imgSrc}
                    onClick={() => this.onGifClick()}/>
                <GifDetails {...{rating, username, showGifDetails, import_datetime}}/>
            </div>);
    }
}