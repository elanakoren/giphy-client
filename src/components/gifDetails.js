// @flow
import React, {Component} from 'react';
import moment from 'moment';

type Props = {
    rating: string,
    username: string,
    import_datetime: string,
    showGifDetails: 'show-details' | 'hide-details',
}

export default class GifDetails extends Component<Props> {
    props: Props;

    render() {
        const {rating, username, showGifDetails, import_datetime} = this.props;
        return <div className={`${showGifDetails} gif-details`}>
            {rating && <div className="gif-detail-text"> rating: {rating} </div>}
            {username && <div className="gif-detail-text"> username: {username} </div>}
            {import_datetime && <div className="gif-detail-text date">uploaded: {moment(import_datetime).format("MMM Do YYYY")}</div>}
            {!rating && !username && !import_datetime && <div>
                No details available
            </div>}
        </div>
    }
}