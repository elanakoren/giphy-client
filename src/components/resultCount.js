// @flow
import React, {Component} from 'react';

type Props = {
    resultCount: ?number
}

export class ResultCount extends Component<Props> {
    props: Props;

    render() {
        const {resultCount} = this.props;
        return (
            <div className="results-count">
                <span className="result">{`${resultCount || 0} result(s) found`}</span>
            </div>
        );
    }
}