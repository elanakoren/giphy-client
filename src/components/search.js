// @flow
import React, {Component} from 'react';

type Props = {
    gifActions: Object
}

export default class Search extends Component<Props> {
    searchQuery: ?HTMLInputElement | null;

    constructor(props: Props) {
        super(props);
        this.searchQuery = null;
    }

    clearSearch() {
        const {gifActions} = this.props;
        // flow demands these type checks
        if (this.searchQuery instanceof HTMLInputElement) {
            // clear the input box
            this.searchQuery.value = '';
        }
        // fetch a new batch of trending gifs to replace the ones on the state
        gifActions.fetchGifs();
    }

    performSearch() {
        const {gifActions} = this.props;
        // Don't hit Giphy with an empty search. Checking that this.searchQuery exists before
        // accessing its value is another demand imposed by flow.
        if (this.searchQuery && !this.searchQuery.value) return;
        else if (this.searchQuery) {
            gifActions.fetchGifs(this.searchQuery.value)
        }

    }

    render() {
        return (
            <div className="search-control">
                <input
                    type="text"
                    ref={(input) => {
                        this.searchQuery = input
                    }}/>
                <button
                    className="search-button submit-button"
                    onClick={() => this.performSearch()}>
                    Search
                </button>
                <button
                    className="search-button clear-button"
                    onClick={() => this.clearSearch()}>
                    Clear Search
                </button>
            </div>
        )
    }
}