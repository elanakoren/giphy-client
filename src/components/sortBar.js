// @flow
import React, {Component} from 'react';
import type {sortOrder} from '../types';

type Props = {
    gifActions: Object,
};

type State = {
    sortOrder: sortOrder,
};

export class SortBar extends Component<Props, State> {
    props: Props;
    state: State;

    constructor(props: Props) {
        super(props);
        this.state = {
            sortOrder: 'desc'
        }
    }

    changeSortOrder() {
        const {gifActions} = this.props;
        const {sortOrder} = this.state;
        let newSortOrder = sortOrder === 'asc' ? 'desc' : 'asc';

        this.setState({sortOrder: newSortOrder}, () => {
            gifActions.changeGifOrder(newSortOrder);
        });
    }

    render() {
        return (
            <div className="sort-control">
                <label
                    htmlFor="select-sort"
                    className="select-label">
                    Sort by upload time:
                </label>
                <select
                    onChange={() => this.changeSortOrder()}
                    id="select-sort"
                    className="sort-dropdown">
                    <option>Newest first</option>
                    <option>Oldest first</option>
                </select>
            </div>
        )
    }
}