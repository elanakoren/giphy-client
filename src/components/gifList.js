import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Gif} from './gif';

// TODO: could be functional
export class GifList extends Component {
    static propTypes = {
        gifData: PropTypes.array,
    };

    render() {
        const {gifData} = this.props;
        const gifs = gifData.map((gifDatum, i) => {
            return (<Gif
                key={i}
                imgSrc={gifDatum.images.downsized.url}
            />)
        });
        return (
            <div>
                {gifs}
            </div>
        );
    }
}