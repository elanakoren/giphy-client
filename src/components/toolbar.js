import React, {Component} from 'react';
import {SortBar} from './sortBar';
import Search from './search';

type Props = {
    gifActions: Object,
    options: Object[],
    label: string,
};


export default class Toolbar extends Component<Props> {
    props: Props;

    render() {
        const {options, gifActions, label} = this.props;
        return (
            <div className="toolbar-grid">
                <Search {...{gifActions}}/>
                <SortBar {...{options, gifActions, label}}/>
            </div>
        )
    }
}