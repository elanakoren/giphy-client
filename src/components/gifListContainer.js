//@flow
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as gifActions from '../actions/actions';
import React, {Component} from 'react';
import {GifList} from './gifList';
import Toolbar from './toolbar';
import type {gifData} from '../types';
import {sortByDateAscending, sortByDateDescending} from '../helpers/sortHelper';

type Props = {
    gifActions: Object,
    gifs: gifData
};

/**
 * This is a connected component: it gets the redux state and
 * the action creators so that they can become props
 * for the downstream components. It also gets the
 * sort state (ascending or descending) from the store
 * and sorts the data, so that GifList can display it in the correct order.
 */
export class GifListContainer extends Component<Props> {

    // perform the initial data fetch
    componentWillMount() {
        this.props.gifActions.fetchGifs()
    }

    render() {
        const {gifs, gifActions} = this.props;
        let gifList;

        const gifDataAvailable = (gifs.data && gifs.data.length > 0);
        let resultCount;

        if (gifDataAvailable && gifs.order === 'asc') {
            gifs.data.sort(sortByDateAscending);
            gifList = (<GifList gifData={gifs.data}/>);
            resultCount = gifs.data.length;
        }

        else if (gifDataAvailable) {
            gifs.data.sort(sortByDateDescending);
            gifList = (<GifList gifData={gifs.data}/>);
            resultCount = gifs.data.length;
        }

        return (
            <div>
                <Toolbar {...{gifActions, resultCount}}/>
                {gifList}
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        gifs: state.gifs
    };
}

function mapDispatchToProps(dispatch) {
    return {
        gifActions: bindActionCreators(gifActions, dispatch)
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(GifListContainer);