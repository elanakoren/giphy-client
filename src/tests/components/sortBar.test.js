//@flow
import React from 'react';
import {SortBar} from '../../components/sortBar';
import {shallow} from 'enzyme';

describe('sortbar component', () => {
    let wrapper, changeGifOrder;
    beforeEach(() =>{
        changeGifOrder = jest.fn();
        wrapper = shallow(<SortBar {...{
            gifActions: {changeGifOrder},
        }}/>);
    });

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('defaults to sorting descendingly (newest first)', () => {
        expect(wrapper.state().sortOrder).toEqual('desc');
    });

    describe('clicking the sortbar', () => {
        beforeEach(() => {
            wrapper.find('select').simulate('change', {target: { value : 'asc'}});
        });

        it('calls the changeGifOrder action and updates the state', () => {
            expect(changeGifOrder).toHaveBeenCalledWith('asc');
            expect(wrapper.state().sortOrder).toEqual('asc');
        });
    });
});