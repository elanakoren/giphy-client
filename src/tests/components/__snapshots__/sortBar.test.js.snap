// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortbar component renders correctly 1`] = `
<div
  className="sort-control"
>
  <label
    className="select-label"
    htmlFor="select-sort"
  >
    Sort by upload time:
  </label>
  <select
    className="sort-dropdown"
    id="select-sort"
    onChange={[Function]}
  >
    <option>
      Newest first
    </option>
    <option>
      Oldest first
    </option>
  </select>
</div>
`;
